src/cc/lib/graph/partition.cc

```diff
@@ -908,16 +908,38 @@ std::optional<size_t> Partition::EdgeFeatureOffset(uint64_t internal_src_node_id
     {
         return std::nullopt;
     }
+
     const auto tp_count = m_edge_type_offset[type_offset + 1] - m_edge_type_offset[type_offset];
     auto fst = std::begin(m_edge_destination) + m_edge_type_offset[type_offset];
     auto lst = fst + tp_count;
-    auto it = std::lower_bound(fst, lst, input_edge_dst);
-    if (it == lst)
+
+    // Fast path: linear search outperforms binary search for small number of elements.
+    if (tp_count <= 16)
     {
-        // Edge was not found in this partition.
+        auto it = std::find(fst, lst, input_edge_dst);
+        if (it != lst && *it == input_edge_dst)
+        {
+            return {it - std::begin(m_edge_destination)};
+        }
+
         return std::nullopt;
     }
-    return {it - std::begin(m_edge_destination)};
+
+    auto it = std::lower_bound(fst, lst, input_edge_dst);
+    if (it != lst && *it == input_edge_dst)
+    {
+        return {it - std::begin(m_edge_destination)};
+    }
+
+    // For temporal graphs we need to use fallback to linear search, since edges are stored based on timestamps first
+    // order.
+    it = std::find(fst, lst, input_edge_dst);
```

Comment:
Can we at least filter input_edge_dst by active timestamps only so we reduce linear search time?

Reply:
If you mean filter and create a new vector, then it is an allocation, which is slow. Without an extra variable it will be a very complex binary search.

Reply:
This approach doesn’t require binary data regeneration for non temporal graphs and is backward compatible.
We can write a complex binary search for temporal graphs with edge features when it becomes a bottleneck.